include/turtle/mock.hpp

```diff
@@ -11,217 +11,106 @@
-        return f(context, instance);                                                                   \
-    }
+/// generates only a non-const conversion operator
+#define MOCK_NON_CONST_CONVERSION_OPERATOR(M, T, identifier) \
+    M T() { return MOCK_ANONYMOUS_HELPER(identifier)(); }    \
+    MOCK_METHOD_HELPER(T(), identifier)
 
-#define MOCK_CONSTRUCTOR_AUX(T, n, A, t, tpn)                                                    \
-    T(MOCK_DECL_PARAMS(n, void A, tpn)) { MOCK_HELPER(t)(MOCK_FORWARD_PARAMS(n, void A, tpn)); } \
-    MOCK_FUNCTION_HELPER(void A, t, static, tpn)
-
-/// MOCK_CONSTRUCTOR( [calling convention] name, arity, parameters, identifier )
-/// As constructors do not have a return type, the usual signature gets restricted here to just the parameters.
-#define MOCK_CONSTRUCTOR(T, n, A, t) MOCK_CONSTRUCTOR_AUX(T, n, A, t, )
 /// MOCK_CONSTRUCTOR( [calling convention] name, arity, parameters, identifier )
-/// must be used if the signature uses a template parameter of the class
 /// As constructors do not have a return type, the usual signature gets restricted here to just the parameters.
-#define MOCK_CONSTRUCTOR_TPL(T, n, A, t) MOCK_CONSTRUCTOR_AUX(T, n, A, t, typename)
+#define MOCK_CONSTRUCTOR(T, arity, parameters, identifier) MOCK_CONSTRUCTOR_AUX(T, arity, parameters, identifier)
 
 /// MOCK_DESTRUCTOR( [calling convention] ~name, identifier )
-#define MOCK_DESTRUCTOR(T, t)           \
-    T()                                 \
-    {                                   \
-        try                             \
-        {                               \
-            MOCK_ANONYMOUS_HELPER(t)(); \
-        } catch(...)                    \
-        {}                              \
-    }                                   \
-    MOCK_METHOD_HELPER(void(), t)
-
-#define MOCK_FUNCTION_AUX(F, n, S, t, s, tpn)                                                             \
-    MOCK_FUNCTION_HELPER(S, t, s, tpn)                                                                    \
-    s MOCK_DECL(F, n, S, , tpn)                                                                           \
-    {                                                                                                     \
-        static_assert(n == mock::detail::function_arity<MOCK_FUNCTION_TYPE(S)>::value, "Arity mismatch"); \
-        return MOCK_HELPER(t)(MOCK_FORWARD_PARAMS(n, S, tpn));                                            \
-    }
-
-#define MOCK_VARIADIC_ELEM_0(e0, ...) e0
-#define MOCK_VARIADIC_ELEM_1(e0, e1, ...) e1
-#define MOCK_VARIADIC_ELEM_2(e0, e1, e2, ...) e2
+#define MOCK_DESTRUCTOR(T, identifier)           \
+    T()                                          \
+    {                                            \
+        try                                      \
+        {                                        \
+            MOCK_ANONYMOUS_HELPER(identifier)(); \
+        } catch(...)                             \
```

Comment:
Should we really be catching an exception thrown in the destructor?

Reply:
As not doing so will terminate the program, I'd say yes.